filebeat/input/tcp/harvester.go

```diff
@@ -0,0 +1,157 @@
+package tcp
+
+import (
+	"bufio"
+	"bytes"
+	"net"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/elastic/beats/filebeat/harvester"
+)
+
+// Harvester represent a TCP server
+type Harvester struct {
+	sync.RWMutex
+	forwarder *harvester.Forwarder
+	config    config
+	server    net.Listener
+	clients   map[*Client]struct{}
+	wg        sync.WaitGroup
+	done      chan struct{}
+	splitFunc bufio.SplitFunc
+}
+
+// NewHarvester creates a new harvester that will forward events
+func NewHarvester(
+	forwarder *harvester.Forwarder,
+	cfg *common.Config,
+) (*Harvester, error) {
+
+	config := defaultConfig
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	server, err := net.Listen("tcp", config.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	sf := splitFunc([]byte(config.LineDelimiter))
+	return &Harvester{
+		config:    config,
+		forwarder: forwarder,
+		clients:   make(map[*Client]struct{}, 0),
+		done:      make(chan struct{}),
+		server:    server,
+		splitFunc: sf,
+	}, nil
+}
+
+// Run start and run a new TCP listener to receive new data
+func (h *Harvester) Run() error {
+	logp.Info("Started listening for incoming TCP connection on: %s:", h.config.Host)
+	for {
+		conn, err := h.server.Accept()
+		if err != nil {
+			select {
+			case <-h.done:
+				return nil
+			default:
+				logp.Debug("tcp", "Can not accept the connection", err)
+				continue
+			}
+		}
+
+		client := NewClient(
+			conn,
+			h.forwarder,
+			h.splitFunc,
+			h.config.MaxMessageSize,
+			h.config.Timeout,
+		)
+		logp.Debug(
+			"tcp",
+			"New client, remote: %s (total clients: %d)",
+			conn.RemoteAddr(),
+			h.clientsCount(),
+		)
+		h.wg.Add(1)
+		go func() {
+			defer conn.Close()
```

Comment:
Seeing that you also close the connection here, I wonder what happens if `Close` is called twice?